### PR TITLE
2400: Avoid overwriting mbox archive

### PR DESCRIPTION
**Pull request by zhaosongzs**
Sometimes, the ArchiveWorkItem fails to retrieve the current contents of an mbox file while processing a PR. When this happens, it will resend all emails. Erik pointed out that in GitLab, creating a new file uses a POST request, while updating an existing file uses PUT. Similarly, GitHub requires the current "SHA" to update a file.

The ArchiveWorkItem can determine whether it found existing content in the mbox archive and choose the right method accordingly. If it incorrectly fails to retrieve the existing content from GitLab or GitHub, it will attempt to create a new file. This attempt should fail, triggering a retry of the WorkItem, which would ideally succeed in retrieving the existing content.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace

### Issue
 * [SKARA-2400](https://bugs.openjdk.org/browse/SKARA-2400): Avoid overwriting mbox archive (**Bug** - P3)

### Reviewers
 * [Erik Joelsson](https://openjdk.org/census#erikj) (@erikj79 - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/skara.git pull/1696/head:pull/1696` \
`$ git checkout pull/1696`

Update a local copy of the PR: \
`$ git checkout pull/1696` \
`$ git pull https://git.openjdk.org/skara.git pull/1696/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1696`

View PR using the GUI difftool: \
`$ git pr show -t 1696`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/skara/pull/1696.diff">https://git.openjdk.org/skara/pull/1696.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/skara/pull/1696#issuecomment-2484225353)
</details>

### CODE REVIEW

File: forge/src/main/java/org/openjdk/skara/forge/HostedRepository.java [hunk 1]
```diff
@@ -108,7 +108,7 @@ PullRequest createPullRequest(HostedRepository target,
      * @param authorName  Name of author and committer for commit
      * @param authorEmail Email of author and committer for commit
      */
-    void writeFileContents(String filename, String content, Branch branch, String message, String authorName, String authorEmail);
+    void writeFileContents(String filename, String content, Branch branch, String message, String authorName, String authorEmail, boolean createNewFile);
```
Comment (erikj79) [inline forge/src/main/java/org/openjdk/skara/forge/HostedRepository.java:111]: Please update the javadoc with the new parameter and make it clear what should happen when set to true or false. If set to true, the call will fail if the file exists. If set to false, the call will fail if the file does not exist.

File: forge/src/main/java/org/openjdk/skara/forge/github/GitHubRepository.java [hunk 2]
```diff
@@ -291,7 +291,7 @@ public Optional<String> fileContents(String filename, String ref) {
     }
 
     @Override
-    public void writeFileContents(String filename, String content, Branch branch, String message, String authorName, String authorEmail) {
+    public void writeFileContents(String filename, String content, Branch branch, String message, String authorName, String authorEmail, boolean createNewEmail) {
```
Comment (erikj79) [inline forge/src/main/java/org/openjdk/skara/forge/github/GitHubRepository.java:294]: This variable name doesn't make sense here.

File: test/src/main/java/org/openjdk/skara/test/TestHostedRepository.java [hunk 3]
```diff
@@ -229,7 +229,7 @@ public Optional<String> fileContents(String filename, String ref) {
     }
 
     @Override
-    public void writeFileContents(String filename, String content, Branch branch, String message, String authorName, String authorEmail) {
+    public void writeFileContents(String filename, String content, Branch branch, String message, String authorName, String authorEmail, boolean createNewFile) {
```
Comment (erikj79) [inline test/src/main/java/org/openjdk/skara/test/TestHostedRepository.java:232]: This implementation should also enforce the new behavior so we catch any mistakes in testing.

File: forge/src/main/java/org/openjdk/skara/forge/HostedRepository.java (hunk 1 above)
Comment (zhaosongzs) [reply forge/src/main/java/org/openjdk/skara/forge/HostedRepository.java:116]: Will do

File: forge/src/main/java/org/openjdk/skara/forge/github/GitHubRepository.java (hunk 2 above)
Comment (zhaosongzs) [reply forge/src/main/java/org/openjdk/skara/forge/github/GitHubRepository.java]: Thanks for catching it! I was in a hurry when working on it

File: test/src/main/java/org/openjdk/skara/test/TestHostedRepository.java (hunk 3 above)
Comment (zhaosongzs) [reply test/src/main/java/org/openjdk/skara/test/TestHostedRepository.java:231]: Will fix it

**Review by erikj79 — APPROVED** [review]

**Comment by zhaosongzs** [comment]
Thank you for the review!
/integrate

**Pull request closed without merging**